[Inductor] Add Nvidia Universal Gemm Backend to Inductor

NikhilAPatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #171205
* #170906
* __->__ #170623


Builds the scaffolding for a full Inductor backend based on NVIDIA’s new [Python universal GEMM template](https://github.com/NVIDIA/cutlass/tree/cutlass_api/python/cutlass_api), including initial setup of the `mm` execution path.

Can benchmark on trunk Tritonbench with:
`clear; TORCHINDUCTOR_FORCE_DISABLE_CACHES=1 TRITON_ALWAYS_COMPILE=1 TRITON_PRINT_AUTOTUNING=1 TORCH_LOGS=+inductor CUDA_VISIBLE_DEVICES=7 python run.py --op gemm --only aten_matmul,pt2_triton_matmul,pt2_nv_universal_gemm_matmul  --metrics tflops,accuracy --force --cudagraph --num-inputs 1
`

Unit tests: 
`test/inductor/test_nv_universal_gemm.py`

Followup PRs planned:
- [x] nvMatmulHeuristics to narrow search space
- [ ] Autotune compilation in parallel
- [ ] Epilogue fusions
- [ ] BMM
- [ ] AddMM
- [ ] Dynamic shape support
- [ ] AOTI support
- [x] Cutlass API kernels requiring a device workspace

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @jananisriram @anshul-si

NikhilAPatel committed:
Add Cutlass Universal Gemm API Gemm to Inductor

[ghstack-poisoned]

NikhilAPatel committed:
Update on "Add Cutlass Universal Gemm API Gemm to Inductor"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

NikhilAPatel committed:
Update on "Add Cutlass Universal Gemm API Gemm to Inductor"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

NikhilAPatel committed:
Update on "Add Cutlass Universal Gemm API Gemm to Inductor"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

NikhilAPatel committed:
Update on "Add Cutlass Universal Gemm API Gemm to Inductor"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

NikhilAPatel committed:
Update on "Add Cutlass Universal Gemm API Gemm to Inductor"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

NikhilAPatel committed:
Update on "Add Cutlass Universal Gemm API Gemm to Inductor"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

NikhilAPatel committed:
Update on "Add Cutlass Universal Gemm API Gemm to Inductor"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

NikhilAPatel committed:
Update on "Add Cutlass Universal Gemm API Gemm to Inductor"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

NikhilAPatel committed:
Update on "Add Cutlass Universal Gemm API Gemm to Inductor"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

NikhilAPatel committed:
Update on "Add Cutlass Universal Gemm API Gemm to Inductor"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

NikhilAPatel committed:
Update on "Add Cutlass Universal Gemm API Gemm to Inductor"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

NikhilAPatel committed:
Update on "[Inductor] Add CuTeDSL Backend for GEMM"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

NikhilAPatel committed:
Update on "[Inductor] Add CuTeDSL Backend for GEMM"



Builds the scaffolding for a full Inductor backend based on NVIDIA’s new [Python universal GEMM template](https://github.com/NVIDIA/cutlass/tree/cutlass_api/python/cutlass_api), including initial setup of the `mm` execution path.

Can benchmark on trunk Tritonbench with:
`clear; TORCHINDUCTOR_FORCE_DISABLE_CACHES=1 TRITON_ALWAYS_COMPILE=1 TRITON_PRINT_AUTOTUNING=1 TORCH_LOGS=+inductor CUDA_VISIBLE_DEVICES=7 python run.py --op gemm --only aten_matmul,pt2_triton_matmul,pt2_cutlass_api_matmul  --metrics tflops,accuracy --force --cudagraph --num-inputs 5
`

Unit tests: 
`test/inductor/test_cutlass_api_gemm.py`

Followup PRs planned:
- [ ] nvMatmulHeuristics to narrow search space
- [ ] Autotune compilation in parallel
- [ ] Epilogue fusions
- [ ] BMM

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo jananisriram anshul-si

[ghstack-poisoned]

NikhilAPatel committed:
Update on "[Inductor] Add CuTeDSL Backend for GEMM"



Builds the scaffolding for a full Inductor backend based on NVIDIA’s new [Python universal GEMM template](https://github.com/NVIDIA/cutlass/tree/cutlass_api/python/cutlass_api), including initial setup of the `mm` execution path.

Can benchmark on trunk Tritonbench with:
`clear; TORCHINDUCTOR_FORCE_DISABLE_CACHES=1 TRITON_ALWAYS_COMPILE=1 TRITON_PRINT_AUTOTUNING=1 TORCH_LOGS=+inductor CUDA_VISIBLE_DEVICES=7 python run.py --op gemm --only aten_matmul,pt2_triton_matmul,pt2_cutlass_api_matmul  --metrics tflops,accuracy --force --cudagraph --num-inputs 5
`

Unit tests: 
`test/inductor/test_cutlass_api_gemm.py`

Followup PRs planned:
- [ ] nvMatmulHeuristics to narrow search space
- [ ] Autotune compilation in parallel
- [ ] Epilogue fusions
- [ ] BMM

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo jananisriram anshul-si

[ghstack-poisoned]

NikhilAPatel committed:
Update on "[Inductor] Add CuTeDSL Backend for GEMM"



Builds the scaffolding for a full Inductor backend based on NVIDIA’s new [Python universal GEMM template](https://github.com/NVIDIA/cutlass/tree/cutlass_api/python/cutlass_api), including initial setup of the `mm` execution path.

Can benchmark on trunk Tritonbench with:
`clear; TORCHINDUCTOR_FORCE_DISABLE_CACHES=1 TRITON_ALWAYS_COMPILE=1 TRITON_PRINT_AUTOTUNING=1 TORCH_LOGS=+inductor CUDA_VISIBLE_DEVICES=7 python run.py --op gemm --only aten_matmul,pt2_triton_matmul,pt2_cutlass_api_matmul  --metrics tflops,accuracy --force --cudagraph --num-inputs 5
`

Unit tests: 
`test/inductor/test_cutlass_api_gemm.py`

Followup PRs planned:
- [ ] nvMatmulHeuristics to narrow search space
- [ ] Autotune compilation in parallel
- [ ] Epilogue fusions
- [ ] BMM

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo jananisriram anshul-si

[ghstack-poisoned]

NikhilAPatel committed:
Update on "[Inductor] Add CuTeDSL Backend for GEMM"



Builds the scaffolding for a full Inductor backend based on NVIDIA’s new [Python universal GEMM template](https://github.com/NVIDIA/cutlass/tree/cutlass_api/python/cutlass_api), including initial setup of the `mm` execution path.

Can benchmark on trunk Tritonbench with:
`clear; TORCHINDUCTOR_FORCE_DISABLE_CACHES=1 TRITON_ALWAYS_COMPILE=1 TRITON_PRINT_AUTOTUNING=1 TORCH_LOGS=+inductor CUDA_VISIBLE_DEVICES=7 python run.py --op gemm --only aten_matmul,pt2_triton_matmul,pt2_cutlass_api_matmul  --metrics tflops,accuracy --force --cudagraph --num-inputs 5
`

Unit tests: 
`test/inductor/test_cutlass_api_gemm.py`

Followup PRs planned:
- [ ] nvMatmulHeuristics to narrow search space
- [ ] Autotune compilation in parallel
- [ ] Epilogue fusions
- [ ] BMM

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo jananisriram anshul-si

[ghstack-poisoned]

NikhilAPatel committed:
Update on "[Inductor] Add Nvidia Universal Gemm Backend to Inductor"



Builds the scaffolding for a full Inductor backend based on NVIDIA’s new [Python universal GEMM template](https://github.com/NVIDIA/cutlass/tree/cutlass_api/python/cutlass_api), including initial setup of the `mm` execution path.

Can benchmark on trunk Tritonbench with:
`clear; TORCHINDUCTOR_FORCE_DISABLE_CACHES=1 TRITON_ALWAYS_COMPILE=1 TRITON_PRINT_AUTOTUNING=1 TORCH_LOGS=+inductor CUDA_VISIBLE_DEVICES=7 python run.py --op gemm --only aten_matmul,pt2_triton_matmul,pt2_nv_universal_gemm_matmul  --metrics tflops,accuracy --force --cudagraph --num-inputs 1
`

Unit tests: 
`test/inductor/test_cutlass_api_gemm.py`

Followup PRs planned:
- [ ] nvMatmulHeuristics to narrow search space
- [ ] Autotune compilation in parallel
- [ ] Epilogue fusions
- [ ] BMM
- [ ] AddMM
- [ ] Dynamic shape support
- [ ] AOTI support
- [ ] Cutlass API kernels requiring a device workspace

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo jananisriram anshul-si

[ghstack-poisoned]

NikhilAPatel committed:
Update on "[Inductor] Add Nvidia Universal Gemm Backend to Inductor"



Builds the scaffolding for a full Inductor backend based on NVIDIA’s new [Python universal GEMM template](https://github.com/NVIDIA/cutlass/tree/cutlass_api/python/cutlass_api), including initial setup of the `mm` execution path.

Can benchmark on trunk Tritonbench with:
`clear; TORCHINDUCTOR_FORCE_DISABLE_CACHES=1 TRITON_ALWAYS_COMPILE=1 TRITON_PRINT_AUTOTUNING=1 TORCH_LOGS=+inductor CUDA_VISIBLE_DEVICES=7 python run.py --op gemm --only aten_matmul,pt2_triton_matmul,pt2_nv_universal_gemm_matmul  --metrics tflops,accuracy --force --cudagraph --num-inputs 1
`

Unit tests: 
`test/inductor/test_cutlass_api_gemm.py`

Followup PRs planned:
- [ ] nvMatmulHeuristics to narrow search space
- [ ] Autotune compilation in parallel
- [ ] Epilogue fusions
- [ ] BMM
- [ ] AddMM
- [ ] Dynamic shape support
- [ ] AOTI support
- [ ] Cutlass API kernels requiring a device workspace

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo jananisriram anshul-si

[ghstack-poisoned]

drisspg reviewed (commented):
Can you add to https://github.com/pytorch/pytorch/blob/a668e3bffacae2157d2f401138f09d6e1ef7ab8c/.ci/pytorch/test.sh#L345-L361

NikhilAPatel committed:
Update on "[Inductor] Add Nvidia Universal Gemm Backend to Inductor"



Builds the scaffolding for a full Inductor backend based on NVIDIA’s new [Python universal GEMM template](https://github.com/NVIDIA/cutlass/tree/cutlass_api/python/cutlass_api), including initial setup of the `mm` execution path.

Can benchmark on trunk Tritonbench with:
`clear; TORCHINDUCTOR_FORCE_DISABLE_CACHES=1 TRITON_ALWAYS_COMPILE=1 TRITON_PRINT_AUTOTUNING=1 TORCH_LOGS=+inductor CUDA_VISIBLE_DEVICES=7 python run.py --op gemm --only aten_matmul,pt2_triton_matmul,pt2_nv_universal_gemm_matmul  --metrics tflops,accuracy --force --cudagraph --num-inputs 1
`

Unit tests: 
`test/inductor/test_nv_universal_gemm.py`

Followup PRs planned:
- [ ] nvMatmulHeuristics to narrow search space
- [ ] Autotune compilation in parallel
- [ ] Epilogue fusions
- [ ] BMM
- [ ] AddMM
- [ ] Dynamic shape support
- [ ] AOTI support
- [ ] Cutlass API kernels requiring a device workspace

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo jananisriram anshul-si

[ghstack-poisoned]

NikhilAPatel committed:
Update on "[Inductor] Add Nvidia Universal Gemm Backend to Inductor"



Builds the scaffolding for a full Inductor backend based on NVIDIA’s new [Python universal GEMM template](https://github.com/NVIDIA/cutlass/tree/cutlass_api/python/cutlass_api), including initial setup of the `mm` execution path.

Can benchmark on trunk Tritonbench with:
`clear; TORCHINDUCTOR_FORCE_DISABLE_CACHES=1 TRITON_ALWAYS_COMPILE=1 TRITON_PRINT_AUTOTUNING=1 TORCH_LOGS=+inductor CUDA_VISIBLE_DEVICES=7 python run.py --op gemm --only aten_matmul,pt2_triton_matmul,pt2_nv_universal_gemm_matmul  --metrics tflops,accuracy --force --cudagraph --num-inputs 1
`

Unit tests: 
`test/inductor/test_nv_universal_gemm.py`

Followup PRs planned:
- [ ] nvMatmulHeuristics to narrow search space
- [ ] Autotune compilation in parallel
- [ ] Epilogue fusions
- [ ] BMM
- [ ] AddMM
- [ ] Dynamic shape support
- [ ] AOTI support
- [ ] Cutlass API kernels requiring a device workspace

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo jananisriram anshul-si

[ghstack-poisoned]

NikhilAPatel committed:
Update on "[Inductor] Add Nvidia Universal Gemm Backend to Inductor"



Builds the scaffolding for a full Inductor backend based on NVIDIA’s new [Python universal GEMM template](https://github.com/NVIDIA/cutlass/tree/cutlass_api/python/cutlass_api), including initial setup of the `mm` execution path.

Can benchmark on trunk Tritonbench with:
`clear; TORCHINDUCTOR_FORCE_DISABLE_CACHES=1 TRITON_ALWAYS_COMPILE=1 TRITON_PRINT_AUTOTUNING=1 TORCH_LOGS=+inductor CUDA_VISIBLE_DEVICES=7 python run.py --op gemm --only aten_matmul,pt2_triton_matmul,pt2_nv_universal_gemm_matmul  --metrics tflops,accuracy --force --cudagraph --num-inputs 1
`

Unit tests: 
`test/inductor/test_nv_universal_gemm.py`

Followup PRs planned:
- [ ] nvMatmulHeuristics to narrow search space
- [ ] Autotune compilation in parallel
- [ ] Epilogue fusions
- [ ] BMM
- [ ] AddMM
- [ ] Dynamic shape support
- [ ] AOTI support
- [ ] Cutlass API kernels requiring a device workspace

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo jananisriram anshul-si

[ghstack-poisoned]

NikhilAPatel committed:
Update on "[Inductor] Add Nvidia Universal Gemm Backend to Inductor"



Builds the scaffolding for a full Inductor backend based on NVIDIA’s new [Python universal GEMM template](https://github.com/NVIDIA/cutlass/tree/cutlass_api/python/cutlass_api), including initial setup of the `mm` execution path.

Can benchmark on trunk Tritonbench with:
`clear; TORCHINDUCTOR_FORCE_DISABLE_CACHES=1 TRITON_ALWAYS_COMPILE=1 TRITON_PRINT_AUTOTUNING=1 TORCH_LOGS=+inductor CUDA_VISIBLE_DEVICES=7 python run.py --op gemm --only aten_matmul,pt2_triton_matmul,pt2_nv_universal_gemm_matmul  --metrics tflops,accuracy --force --cudagraph --num-inputs 1
`

Unit tests: 
`test/inductor/test_nv_universal_gemm.py`

Followup PRs planned:
- [ ] nvMatmulHeuristics to narrow search space
- [ ] Autotune compilation in parallel
- [ ] Epilogue fusions
- [ ] BMM
- [ ] AddMM
- [ ] Dynamic shape support
- [ ] AOTI support
- [ ] Cutlass API kernels requiring a device workspace

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo jananisriram anshul-si

[ghstack-poisoned]

NikhilAPatel committed:
Update on "[Inductor] Add Nvidia Universal Gemm Backend to Inductor"



Builds the scaffolding for a full Inductor backend based on NVIDIA’s new [Python universal GEMM template](https://github.com/NVIDIA/cutlass/tree/cutlass_api/python/cutlass_api), including initial setup of the `mm` execution path.

Can benchmark on trunk Tritonbench with:
`clear; TORCHINDUCTOR_FORCE_DISABLE_CACHES=1 TRITON_ALWAYS_COMPILE=1 TRITON_PRINT_AUTOTUNING=1 TORCH_LOGS=+inductor CUDA_VISIBLE_DEVICES=7 python run.py --op gemm --only aten_matmul,pt2_triton_matmul,pt2_nv_universal_gemm_matmul  --metrics tflops,accuracy --force --cudagraph --num-inputs 1
`

Unit tests: 
`test/inductor/test_nv_universal_gemm.py`

Followup PRs planned:
- [ ] nvMatmulHeuristics to narrow search space
- [ ] Autotune compilation in parallel
- [ ] Epilogue fusions
- [ ] BMM
- [ ] AddMM
- [ ] Dynamic shape support
- [ ] AOTI support
- [ ] Cutlass API kernels requiring a device workspace

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo jananisriram anshul-si

[ghstack-poisoned]

NikhilAPatel committed:
Update on "[Inductor] Add Nvidia Universal Gemm Backend to Inductor"



Builds the scaffolding for a full Inductor backend based on NVIDIA’s new [Python universal GEMM template](https://github.com/NVIDIA/cutlass/tree/cutlass_api/python/cutlass_api), including initial setup of the `mm` execution path.

Can benchmark on trunk Tritonbench with:
`clear; TORCHINDUCTOR_FORCE_DISABLE_CACHES=1 TRITON_ALWAYS_COMPILE=1 TRITON_PRINT_AUTOTUNING=1 TORCH_LOGS=+inductor CUDA_VISIBLE_DEVICES=7 python run.py --op gemm --only aten_matmul,pt2_triton_matmul,pt2_nv_universal_gemm_matmul  --metrics tflops,accuracy --force --cudagraph --num-inputs 1
`

Unit tests: 
`test/inductor/test_nv_universal_gemm.py`

Followup PRs planned:
- [ ] nvMatmulHeuristics to narrow search space
- [ ] Autotune compilation in parallel
- [ ] Epilogue fusions
- [ ] BMM
- [ ] AddMM
- [ ] Dynamic shape support
- [ ] AOTI support
- [ ] Cutlass API kernels requiring a device workspace

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo jananisriram anshul-si

[ghstack-poisoned]

NikhilAPatel commented:
@pytorchbot merge -i

NikhilAPatel committed:
Update on "[Inductor] Add Nvidia Universal Gemm Backend to Inductor"



Builds the scaffolding for a full Inductor backend based on NVIDIA’s new [Python universal GEMM template](https://github.com/NVIDIA/cutlass/tree/cutlass_api/python/cutlass_api), including initial setup of the `mm` execution path.

Can benchmark on trunk Tritonbench with:
`clear; TORCHINDUCTOR_FORCE_DISABLE_CACHES=1 TRITON_ALWAYS_COMPILE=1 TRITON_PRINT_AUTOTUNING=1 TORCH_LOGS=+inductor CUDA_VISIBLE_DEVICES=7 python run.py --op gemm --only aten_matmul,pt2_triton_matmul,pt2_nv_universal_gemm_matmul  --metrics tflops,accuracy --force --cudagraph --num-inputs 1
`

Unit tests: 
`test/inductor/test_nv_universal_gemm.py`

Followup PRs planned:
- [ ] nvMatmulHeuristics to narrow search space
- [ ] Autotune compilation in parallel
- [ ] Epilogue fusions
- [ ] BMM
- [ ] AddMM
- [ ] Dynamic shape support
- [ ] AOTI support
- [ ] Cutlass API kernels requiring a device workspace

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo jananisriram anshul-si

[ghstack-poisoned]
